### PR TITLE
fix(maskeditor): solid mask coverage in GPU MaskEditor strokes

### PR DESCRIPTION
**Pull request by silveroxides**
## Summary

Fix GPU MaskEditor strokes producing mask values of `254` instead of `255`
with opacity and hardness set to maximum.

The stroke accumulator previously used source-over blending with a fixed
opacity of `0.5`. Overlapping brush samples therefore accumulated opacity
instead of recording brush coverage. GPU readback also truncated normalized
values when converting them to bytes.

The save-path RGB regression is handled separately by #13758 and is no longer
part of this PR.

## Changes

- Accumulate the maximum coverage contributed by brush samples.
- Pass full coverage through both GPU stroke paths.
- Apply user opacity once when compositing the stroke.
- Round GPU readback values when converting them to bytes.
- Add regression tests for maximum blending and full stroke coverage.

## Review Focus

Soft brush edges remain fractional. This change does not clamp values near
`255`; it makes fully covered pixels reach `255` through the normal rendering
path.

The existing `brushOpacity` uniform name is retained. Coverage is distinguished
only at the stroke accumulator boundary to keep the schema and shaders focused.

## Testing

- Focused Vitest: 2 files, 51 tests passed
- `pnpm typecheck`: passed
- Targeted ESLint and oxfmt checks: passed
- Manual test: a fully opaque hard stroke produces mask value `255`

### CODE REVIEW

File: src/composables/maskeditor/useGPUResources.test.ts [hunk 1]
```diff
@@ -184,9 +199,54 @@ describe('initPreviewCanvas', () => {
...
   })
+
+  it('uses full coverage for each GPU stroke sample', async () => {
+    vi.stubGlobal('GPUTextureUsage', {
+      TEXTURE_BINDING: 0x0004,
+      STORAGE_BINDING: 0x0080,
+      RENDER_ATTACHMENT: 0x0010,
+      COPY_DST: 0x0008,
+      COPY_SRC: 0x0001
+    })
```
[the remark this answers is not in the document]
Comment (silveroxides) [reply src/composables/maskeditor/useGPUResources.test.ts:216]: Fixed in 4039dd97f. The test stub now uses the WebGPU-defined values: COPY_DST 0x0002 and STORAGE_BINDING 0x0008.

File: src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts [hunk 2]
```diff
@@ -171,6 +171,27 @@ describe('GPUBrushRenderer', () => {
...
+          dstFactor: 'one',
+          operation: 'max'
+        },
+        alpha: {
+          srcFactor: 'one',
+          dstFactor: 'one',
+          operation: 'max'
+        }
+      })
+    })
```
[the remark this answers is not in the document]
Comment (silveroxides) [reply src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts:193]: These Vitest tests run in happy-dom with a fully mocked GPUDevice, so no WGSL executes and there are no rendered pixels to read back. Simulating pixel output here would only test the mock. The available boundaries are covered separately: the accumulator descriptor asserts max blending, the uniform data asserts coverage 1, and the readback shader rounds before byte conversion. Real GPU output was also manually verified at 255. A true rendered-pixel assertion would require a browser test environment with functional WebGPU, which this test suite does not currently provide.

File: src/composables/maskeditor/useMaskEditorSaver.test.ts [hunk 3]
```diff
@@ -182,23 +197,58 @@ describe('useMaskEditorSaver', () => {
...
+      subfolder: '',
+      type: 'input'
+    })
+
+    const paintedMaskedBody = fetchApiMock.mock.calls[3][1]?.body as FormData
+    expect(JSON.parse(String(paintedMaskedBody.get('original_ref')))).toEqual({
+      filename: 'clipspace-painted-123.png',
+      subfolder: '',
+      type: 'input'
+    })
```
[the remark this answers is not in the document]
Comment (silveroxides) [reply src/composables/maskeditor/useMaskEditorSaver.test.ts]: Fixed in 5ad93e61d. Mock response names now come from each request filename, mask payloads are asserted without order dependence, and only the painted-image-before-painted-mask dependency remains.

File: src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts [hunk 4]
```diff
@@ -171,6 +171,27 @@ describe('GPUBrushRenderer', () => {
       // readback = 1 compute pipeline
       expect(device.createComputePipeline).toHaveBeenCalledTimes(1)
     })
+
+    it('uses maximum blending for accumulated stroke coverage', () => {
+      const accumulatePipeline = (
+        device.createRenderPipeline as ReturnType<typeof vi.fn>
+      ).mock.calls[1][0] as GPURenderPipelineDescriptor
+      const target = accumulatePipeline.fragment as GPUFragmentState
+      const blend = target.targets?.[0]?.blend
```
Comment (DrJKL) [inline src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts]: Use `vi.mocked` for mock call access instead of asserting this is a generic `vi.fn`; that keeps the call typing tied to `createRenderPipeline`.

```suggestion
const accumulatePipeline = vi.mocked(device.createRenderPipeline).mock
  .calls[1][0]
const blend = accumulatePipeline.fragment?.targets?.[0]?.blend
```

File: src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts [hunk 5]
```diff
@@ -242,6 +263,10 @@ describe('GPUBrushRenderer', () => {
       // uniform + instance data = 2 writeBuffer calls
       expect(device.queue.writeBuffer).toHaveBeenCalledTimes(2)
       expect(device.queue.submit).toHaveBeenCalled()
+
+      const uniformData = (device.queue.writeBuffer as ReturnType<typeof vi.fn>)
+        .mock.calls[0][2] as ArrayBuffer
```
Comment (DrJKL) [inline src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts]: Use `vi.mocked` for the mock call access here too. The `ArrayBuffer` assertion is still doing useful narrowing from WebGPU's broader buffer source type.

```suggestion
const uniformData = vi.mocked(device.queue.writeBuffer).mock
  .calls[0][2] as ArrayBuffer
```

**Review by DrJKL — COMMENTED** [review]
Added suggested edits for the applicable mock typing cleanups.

File: src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts (hunk 4 above)
Comment (silveroxides) [reply src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts]: Addressed in 6c7fc74ec. The pipeline call now uses `vi.mocked`; target access remains iterable-safe for the current WebGPU types.

File: src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts (hunk 5 above)
Comment (silveroxides) [reply src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts]: Addressed in 6c7fc74ec. The `writeBuffer` call now uses `vi.mocked` while retaining the `ArrayBuffer` narrowing.

**Comment by jtydhr88** [comment]
The GPU coverage fix is solid, the max-blend accumulator, single opacity application at composite time, and rounded readback correctly get a full-strength stroke to 255, and we've reproduced the 254 symptom locally. 
However, the /upload/mask change (the "preserve RGB under masks" commit) should be dropped from this PR, for 4 reasons:
1. It's now redundant. #13758 (merged) already preserves the source RGB under zero-alpha mask pixels on the client side: the masked and painted-masked layers are encoded straight from the ImageData buffer via encodeRgbaAsPng, so the uploaded PNG already carries the correct RGB. Having the server re-composite it from original_ref does the same work a second time and throws away bytes the client already produced correctly.
2. It reintroduces a deliberately deprecated endpoint. #12318 moved mask compositing off /upload/mask on purpose, and #13758 explicitly declined to restore it for exactly this reason. Going back to /upload/mask + original_ref re-couples the editor to a server contract we're trying to retire, and adds a cross-service dependency the cloud path no longer needs.

3. After rebasing onto current main it conflicts semantically, not textually. The commit applies cleanly (it touches the upload  functions while #13758 touched the encoding functions), but the result is contradictory: the client encodes RGB-preserving PNGs and then asks the server to rebuild them from the original anyway. The accompanying test changes also replace the unified-upload-contract  assertions that just landed on main.

4. the BrushUniforms.brushOpacity → opacity rename churns the schema, both WGSL shaders, and the renderer. The coverage distinction in renderStrokeToAccumulator is well-motivated, but the uniform rename could be reverted to keep the diff focused.

I will open seperate PR for this

**Comment by silveroxides** [comment]
@jtydhr88 The original scope of this PR was the solid-mask GPU issue. The save-path RGB regression was uncovered afterward through manual testing of the fix. I understand that #13758 now supersedes the /upload/mask change and that those commits should be dropped.

What I do not understand is why this PR needed to be closed and replaced by #14198 rather than asking me to rebase and remove the superseded commits.
In addition to automated tests, I built and manually tested the frontend after each change, which is how the existing regression was discovered.
Closing and replacing the PR in this way sends an unfortunate signal that contributors should avoid investigating adjacent regressions found during validation, because doing so may result in their original accepted fix being moved into a maintainer-owned PR.
If #14198 remains the replacement, please explicitly identify #14042 and @silveroxides as the source of the WebGPU fix in its description.

**Comment by jtydhr88** [comment]
@silveroxides sorry, my bad, previously I was thinking I want to make changes directly on your  PR but I could not, because you sent from your fork.
So i just sent PR which I already included your commits.
Let me close mine, and reopen yours

**Comment by silveroxides** [comment]
Thanks. I rebased #14042 onto current main, removed the superseded saver/upload commits, retained the existing brush opacity uniform name, and added coverage for the single-point GPU path. The PR description is now limited to the WebGPU fix.

File: src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts [hunk 6]
```diff
@@ -242,6 +262,10 @@ describe('GPUBrushRenderer', () => {
       // uniform + instance data = 2 writeBuffer calls
       expect(device.queue.writeBuffer).toHaveBeenCalledTimes(2)
       expect(device.queue.submit).toHaveBeenCalled()
+
+      const uniformData = vi.mocked(device.queue.writeBuffer).mock
+        .calls[0][2] as ArrayBuffer
```
Comment (DrJKL) [inline src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts:267]: Shouldn't the types be inferrable?

```suggestion
        .calls[0][2]
```

File: src/composables/maskeditor/gpu/GPUBrushRenderer.ts [hunk 7]
```diff
@@ -463,7 +462,7 @@ export class GPUBrushRenderer {
       this.currentStrokeView,
       this.accumulatePipeline,
       points,
-      settings
+      { ...settings, opacity: settings.coverage }
```
Comment (DrJKL) [inline src/composables/maskeditor/gpu/GPUBrushRenderer.ts:465]: I don't love when one concept gets multiple names in the same file :'(

File: src/composables/maskeditor/useGPUResources.test.ts [hunk 8]
```diff
@@ -8,8 +8,23 @@ vi.mock('typegpu', () => ({
...
 }))
 
+const mockRenderer = vi.hoisted(() => ({
+  clearPreview: vi.fn(),
+  compositeStroke: vi.fn(),
+  destroy: vi.fn(),
+  prepareStroke: vi.fn(),
+  renderStrokeToAccumulator: vi.fn(),
+  blitToCanvas: vi.fn()
+}))
```
Comment (DrJKL) [inline src/composables/maskeditor/useGPUResources.test.ts:19]:
```suggestion
const mockRenderer = vi.hoisted(() => ({
  blitToCanvas: vi.fn(),
  clearPreview: vi.fn(),
  compositeStroke: vi.fn(),
  destroy: vi.fn(),
  prepareStroke: vi.fn(),
  renderStrokeToAccumulator: vi.fn()
}))
```

File: src/composables/maskeditor/useGPUResources.test.ts [hunk 9]
```diff
@@ -52,6 +67,46 @@ function setup() {
   return scope.run(() => useGPUResources())!
 }
 
+async function setupInitializedResources() {
+  vi.stubGlobal('GPUTextureUsage', {
+    TEXTURE_BINDING: 0x0004,
+    STORAGE_BINDING: 0x0008,
+    RENDER_ATTACHMENT: 0x0010,
+    COPY_DST: 0x0002,
+    COPY_SRC: 0x0001
```
Comment (DrJKL) [inline src/composables/maskeditor/useGPUResources.test.ts]:
```suggestion
    COPY_DST: 0x0002,
    COPY_SRC: 0x0001,
    RENDER_ATTACHMENT: 0x0010,
    STORAGE_BINDING: 0x0008,
    TEXTURE_BINDING: 0x0004
```

I should really have a linter do this...

File: src/composables/maskeditor/useGPUResources.test.ts [hunk 10]
```diff
@@ -52,6 +67,46 @@ function setup() {
...
+  })
+  Object.defineProperty(navigator, 'gpu', {
+    configurable: true,
+    value: { getPreferredCanvasFormat: vi.fn(() => 'rgba8unorm') }
+  })
+
+  const imageData = {
+    data: new Uint8ClampedArray(4 * 4 * 4)
+  } as ImageData
```
Comment (DrJKL) [inline src/composables/maskeditor/useGPUResources.test.ts]: Does this work? I like to minimize use of `as`

```suggestion
  const imageData: ImageData = {
    data: new Uint8ClampedArray(4 * 4 * 4)
  }
```

File: src/composables/maskeditor/useGPUResources.test.ts [hunk 11]
```diff
@@ -52,6 +67,46 @@ function setup() {
...
+      createView: vi.fn(),
+      destroy: vi.fn()
+    })),
+    queue: {
+      writeTexture: vi.fn()
+    }
+  }
+  mockStore.tgpuRoot = { device: gpuDevice } as unknown
+  mockStore.maskCanvas = { width: 4, height: 4 } as HTMLCanvasElement
+  mockStore.rgbCanvas = { width: 4, height: 4 } as HTMLCanvasElement
```
Comment (DrJKL) [inline src/composables/maskeditor/useGPUResources.test.ts]: Would shoehorn work here?

File: src/composables/maskeditor/useGPUResources.test.ts (hunk 10 above)
Comment (silveroxides) [reply src/composables/maskeditor/useGPUResources.test.ts]: The suggested object is missing required ImageData fields, so it does not typecheck directly. Replaced the assertion with romPartial<ImageData> in aa4b5e10e.

File: src/composables/maskeditor/useGPUResources.test.ts (hunk 9 above)
Comment (silveroxides) [reply src/composables/maskeditor/useGPUResources.test.ts]: Addressed in aa4b5e10e.

File: src/composables/maskeditor/useGPUResources.test.ts (hunk 11 above)
Comment (silveroxides) [reply src/composables/maskeditor/useGPUResources.test.ts]: Yes. Replaced the partial DOM mock assertions with romPartial and removed the unnecessary root assertion in aa4b5e10e.

File: src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts (hunk 6 above)
Comment (silveroxides) [reply src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts:267]: The assertion is required here. writeBuffer exposes GPUAllowSharedBufferSource, which is not accepted by the Float32Array constructor. Removing it fails typecheck with TS2769. This also matches the earlier review guidance on this line.

File: src/composables/maskeditor/useGPUResources.test.ts (hunk 8 above)
Comment (silveroxides) [reply src/composables/maskeditor/useGPUResources.test.ts:19]: Addressed in aa4b5e10e.

File: src/composables/maskeditor/gpu/GPUBrushRenderer.ts (hunk 7 above)
Comment (silveroxides) [reply src/composables/maskeditor/gpu/GPUBrushRenderer.ts:465]: coverage and opacity are intentionally separate concepts. Coverage controls accumulation for each dab and must be 1 to avoid the 254 output; user opacity is applied once when the accumulated stroke is composited. This call is the boundary where coverage is passed through the shared renderer's opacity-shaped uniform layout.

File: src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts (hunk 6 above)
Comment (DrJKL) [reply src/composables/maskeditor/gpu/GPUBrushRenderer.test.ts:267]: 😢

**Review by jtydhr88 — APPROVED** [review]
